### PR TITLE
Routine maintenance

### PR DESCRIPTION
**Pull request by ericmjl**
@rlouf I added just a few small things.

Primarily a gitignore, and then planning some things like a [pre-commit](https://pre-commit.com/) configuration, and an environment spec file. I'll PR them bit by bit, to make the review easy.

As you might be able to tell, I'm kinda quite interested in mcx. Brought it up to the rest of the PyMC4 dev team on the last call I was on, as I wanted to do some API experiments. JAX seems to be the "right" next thing for building numerical Python packages.

### CODE REVIEW

**Comment by rlouf** [comment]
@ericmjl Thank you, these would be very useful additions. I’ll review later this week.

I’m really humbled by your interest in the project. Would you like to have a quick chat in the coming week so you can share your ideas? I can give you an honest tour of the library (good, bad & ugly), what I’m currently doing and the current plans. Would love to have feedback from someone who has experience working on a PPL that is actually used. We can move this discussion to Twitter DMs from now on.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+var/
+wheels/
+pip-wheel-metadata/
+share/python-wheels/
+*.egg-info/
+.installed.cfg
+*.egg
+MANIFEST
+
+# PyInstaller
```
Comment (rlouf) [inline .gitignore:34]: Do we need PyInstaller? This is a library so I don't think anyone will ever compile it to a binary. Did you have smt else in mind?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+#  Usually these files are written by a python script from a template
+#  before PyInstaller builds the exe, so as to inject date/other infos into it.
+*.manifest
+*.spec
+
+# Installer logs
+pip-log.txt
+pip-delete-this-directory.txt
+
+# Unit test / coverage reports
```
Comment (rlouf) [inline .gitignore:44]: Could we remove anything not related to pytest for now?

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+coverage.xml
+*.cover
+.hypothesis/
+.pytest_cache/
+
+# Translations
+*.mo
+*.pot
+
+# Scrapy stuff:
```
Comment (rlouf) [inline .gitignore:61]: remove?

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+.scrapy
+
+# Sphinx documentation
```
Comment (rlouf) [inline .gitignore:64]: The way the doc currently works is that I build the source in `/docs/build`, commit and upload. Github pages does the rest. So maybe this can go too?

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+docs/_build/
+
+# PyBuilder
```
Comment (rlouf) [inline .gitignore:67]: What is PyBuilder? Is it like PyInstaller?

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+target/
+
+# pyenv
+.python-version
+
+# pipenv
```
Comment (rlouf) [inline .gitignore:73]: You can just delete when not used.

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+#   According to pypa/pipenv#598, it is recommended to include Pipfile.lock in version control.
+#   However, in case of collaboration, if having platform-specific dependencies or dependencies
+#   having no cross-platform support, pipenv may install dependencies that don't work, or not
+#   install all needed dependencies.
+#Pipfile.lock
+
+# celery beat schedule file
```
Comment (rlouf) [inline .gitignore:80]: I don't think we'll ever use Celery.

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+.mypy_cache/
+.dmypy.json
+dmypy.json
+
+# Pyre type checker
+.pyre/
+
+# End of https://www.gitignore.io/api/python
+
+# Custom stuff
```
Comment (rlouf) [inline .gitignore:111]: Can't you just remove every instance of `.ipynb_checkpoints`?

File: .gitignore (hunk 4 above)
Comment (ericmjl) [reply .gitignore:64]: Built artifacts probably are best left to a CI system to build, and then pushed to a `gh-pages` branch; this is a common pattern in many open source projects. I remember when merging the latest master into my fork, I had a ton of conflicts to resolve manually for whatever reason. If you still prefer, I can remove it, but I think encouraging common community practice is better.

File: .gitignore (hunk 1 above)
Comment (ericmjl) [reply .gitignore:34]: This comes from http://gitignore.io/api/python, which covers a wide range of common Python files that never should be added in, based on common community practices. As mentioned in a comment below, I think relying on these opinionated community-built tools to adhere to common community practices could be better, as then we never have to think about them in the future. But if you still disagree, let me know, I'll remove the ones you request to remove.

File: .gitignore (hunk 8 above)
Comment (ericmjl) [reply .gitignore:111]: Sure, good point.

**Comment by talesa** [comment]
>We can move this discussion to Twitter DMs if you’d like.

Hey guys! If you don't mind keeping the discussion public then please do continue here, I'm having a look every now and then :)

**Comment by ericmjl** [comment]
Hey @rlouf!

> I’m really humbled by your interest in the project. 

Please don't elevate me like that! :joy: I am just another PPL user who likes nice APIs :). I'm also not very well-versed in the internals of some of the advanced parts of a PPL, so seeing an API design from scratch is quite educational for me as well.

> Would you like to have a quick chat in the coming week so you can share your ideas? I can give you an honest tour of the library (good, bad & ugly), what I’m currently doing and the current plans. Would love to have feedback from someone who has experience working on a PPL that is actually used. We can move this discussion to Twitter DMs if you’d like.

Yeah, let's do it! I think we should set up a call, and take notes on the issue tracker. Centralizing comms so that others like @talesa can keep tabs is a great way to get feedback from others and possibly build a community of practice around an idea. What do you think?

File: .gitignore (hunk 4 above)
Comment (rlouf) [reply .gitignore:64]: I agree with you, the way it is currently done is very hacky and this is something CI should take care of. If you can do it quickly it would be amazing, otherwise we can do it in another PR and keep this line for now.

File: .gitignore (hunk 1 above)
Comment (rlouf) [reply .gitignore:34]: While I do agree on common practices, I prefer not to add lines that will not be relevant to the project in a foreseeable future. I usually start with a core of known offenders in the context of the library and build from there. I remember once seeing `.tox` in a project's gitignore and spending 5 minutes to make sure that I did not miss anything; tox was indeed not used in the project, the gitignore was just copy-pasted :)

I saw you edited [PyMC4's gitignore](https://github.com/pymc-devs/pymc4/blob/master/.gitignore), it is very well organized. Could we add the ones related to IDEs here? And maybe also `.env` and `.venv`?

**Comment by rlouf** [comment]
> Yeah, let's do it! I think we should set up a call,

Would love to, you can contact me on Twitter to set up a call (I can't DM you).

> and take notes on the issue tracker. Centralizing comms so that others like @talesa can keep tabs is a great way to get feedback from others and possibly build a community of practice around an idea. What do you think?

Great idea! It's also great for us; I've had several discussions already and I don't remember exactly who said what :/ We can use a label to make these notes searchable.

File: .gitignore (hunk 1 above)
Comment (ericmjl) [reply .gitignore:34]: Just added them, will be pushing shortly.

File: .gitignore (hunk 2 above)
Comment (ericmjl) [reply .gitignore:44]: Going to keep coverage and hypothesis. Hypothesis especially - I'd highly recommend using it for testing numerical libraries. It's great, and in some deep learning models that I've implemented, Hypothesis has caught edge cases that I wouldn't have thought of.

File: .gitignore (hunk 4 above)
Comment (ericmjl) [reply .gitignore:64]: Let's do it in another PR.

Do you like Sphinx? I've been test-driving mkdocs recently, and with the right plugins, it's possible to make super cool sites like the following ones:

* http://bayesian-analysis-recipes.ericmjl.com/
* https://fastapi.tiangolo.com/

API docs are possible too, with auto-parsing of docstrings. My experiments with the PyMC4 docs tell me it's doable, and they look gorgeous. But the tradeoff is docstrings have to be written in Markdown, so I wouldn't recommend it for a library that is extensively invested in sphinx-style docstrings. Let me know; I'll keep in mkdocs for now unless you choose otherwise.

File: .gitignore (hunk 5 above)
Comment (ericmjl) [reply .gitignore:67]: No idea! But I'll remove them.

File: .gitignore (hunk 4 above)
Comment (rlouf) [reply .gitignore:53]: I don’t think anyone is a fan of Sphinx, but it does the job quite well. In particular the API docs which I find very important. I don’t know mkdocs very well and I am not downright opposed to it. Let’s see how this works for PyMC4 first :)

**Comment by rlouf** [comment]
The gitignore looks great now, thank you! One hill I am ready to die on though is that code coverage sucks. Avi Bryant recently [tweeted an article](https://twitter.com/avibryant/status/1234223867033292800?s=21) that summarizes my view about tests very well. I’ll probably need to add that to developer resources one day.

So we can just delete these two lines, and then move on to pre-commit which I am really excited about! Hooks are so great yet hard to share. Can’t wait to see how it works.

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+.env
+.venv
+
+# Custom stuff
+*.profraw
+*.DS_Store
+
+# Apparently both patterns _have_ to be present on my system.
+# Having both is a nice backup, I guess.
+*/.ipynb_checkpoints/*
```
Comment (rlouf) [inline .gitignore:113]: You don't need the * here: `.ipynb_checkpoints` is enough.

Comment (ericmjl) [reply .gitignore]: Done.

**Comment by rlouf** [comment]
gitignore looks good to me!

**Comment by rlouf** [comment]
Thank you for the addition @ericmjl ! I rebased the branch on master, merging and closing as we discussed.
